### ISSUE TITLE
`inline` methods can leak references to private classes

### ISSUE THREAD

**Issue by LPTK** [lampepfl/dotty#8194]
## minimized code

```Scala
object Test {
  private class A() { def test = 42 }
  inline def foo: Int = A().test
}
@main def main = Test.foo
```

https://scastie.scala-lang.org/or0Eg6u9Tky1ImaUNEdOug

## Compilation output

```scala
class A in object Test cannot be accessed as a member of (Test : Test.type) from module class main$package$.
```

## expectation

This should compile, or at least report an error at the definition site of the `inline` method (similarly to how errors are reported when classes leak private val references), not at call sites.

**Comment by odersky** [comment]
The fine print of this is actually very tricky. The problem in general is that an inline method could be called in a contect where a type it refers to is not accessible. This looks very hard to implement without getting false negatives. Even the limited case of just flagging accesses to private types looks non-trivial. It requires a complete refactoring of the code in `checkNoPrivateLeaks` which itself is quite tricky. So, this will be time consuming to do, and the benefits are not that great. I think this is a case where outside contributions would be welcome and not much will happen without them.

**Comment by hamzaremmal** [comment]
I'm going over some old issues. Currently, the snippet will error with the following:
```scala
-- [E173] Reference Error: test.scala:5:22 -------------------------------------
5 |@main def main = Test.foo
  |                 ^^^^^^^^
  |class A cannot be accessed as a member of (Test : Test.type) from the top-level definitions in package <empty>.
  |  private class A can only be accessed from object Test.
  |-----------------------------------------------------------------------------
  |Inline stack trace
  |- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
  |This location contains code that was inlined from test.scala:3
3 |  inline def foo: Int = A().test
  |                        ^
   -----------------------------------------------------------------------------
```

I find the error message to contain exactly the information needed (it was inlined, what went wrong during the inlining and where in the inlined method failed).

> This should compile, 

I understand why we would suggest this, and I will disagree because  this means that nothing can be `private` or `protected` in the bytecode because it could be accessed from anywhere via inlining. I personally think that would be a bad design in a language that actually allows interop with JVM languages (since everything will be public in Java/Kotlin too).

> or at least report an error at the definition site of the inline method (similarly to how errors are reported when classes leak private val references), not at call sites

I again respectfully disagree. We could call that inline method in a context where that type is actually accessible and why would I want to restrict that use case (for example in a big codebase). The more expressive option is the one we currently have; we inline, if it works then great if it doesn't, we explain why.

**Comment by LPTK** [comment]
@odersky @hamzaremmal Thanks for the explanations and rationale. I think you make fair points, and it's clear that things are more tricky than it may seem!

IIRC (from over 5 years ago!), my observation was originally motivated by the fact that

```scala
object Test {
  private def A() = 42
  inline def foo: Int = A()
}
@main def main = Test.foo
```

compiles fine, so the treatment of private members just seems somewhat inconsistent, from a user's point of view.

But I agree it may not be a fight worth fighting.

**Comment by hamzaremmal** [comment]
I agree with the fact that this might seem a bit inconsistent but there is a huge difference. In here, we do not leak any private class and we will have no way to have some kind of proxies like for methods. Also, this code will emit the following warning when compiled with `-WunstableInlineAccessors` today.
```scala
-- [E192] Compatibility Warning: test.scala:3:24 -------------------------------
3 |  inline def foo: Int = A()
  |                        ^
  |           Unstable inline accessor inline$A was generated in object Test.
  |
  | longer explanation available when compiling with `-explain`
1 warning found
```
There is also [SIP-52](https://docs.scala-lang.org/sips/52.html) that is designed to make these use cases less inconsistent.
